src/firmware/types.rs

```diff
@@ -202,3 +208,522 @@ pub struct SnpPlatformStatus {
     /// Reported TCB version.
     pub reported_tcb_version: TcbVersion,
 }
+
+#[repr(C, packed)]
+pub struct SnpConfig {
+    reported_tcb: u64,
+    mask_chip_id: u32,
+    reserved: [u8; 52],
+}
+
+impl Default for SnpConfig {
+    fn default() -> Self {
+        Self {
+            reported_tcb: Default::default(),
+            mask_chip_id: Default::default(),
+            reserved: [0; 52],
+        }
+    }
+}
+
+#[repr(C)]
+pub struct SnpSetExtConfig<'a> {
+    /// Address of the SnpConfig or 0 when reported_tcb does not need
+    /// to be updated.
+    pub config_address: Option<&'a SnpConfig>,
+
+    /// Address of extended guest request certificate chain or 0 when
+    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    pub certs_address: Option<*mut c_void>,
```

Comment:
I still need to study Rust pointers in depth, but in the SET command the buffer pointed by certs_address is only read from (and SETs the content in the kernel), so not sure why `mut` is needed here.

Also I see that elsewhere in this project ([snp.rs](https://github.com/virtee/sev/blob/main/src/launch/linux/snp.rs#L115)) we use `u64` as the type for these fields. When we get a `&[u8]` buffer from the caller, we [convert it to address](https://github.com/virtee/sev/blob/main/src/launch/linux/snp.rs#L138) using `addr.as_ptr() as u64`.

Reply:
So, we typically distinguish between types presented to the kernel via ioctl calls, and types that are presented to the user of the library (from Rust). I'll give an example of an ioctl already present in the library.

Let's look at the `SNP_PLATFORM_STATUS` ioctl:

There are two types presented for this ioctl:
1. [SnpStatus](https://github.com/virtee/sev/blob/d52c5c56b04c44a00674f4190e2f3083376af72f/src/firmware/mod.rs#L305): This is the Rust-friendly type that is returned to the user from the [snp_platform_status API](https://github.com/virtee/sev/blob/2e54dddffb3ee7cafc805bca203c1136cf41648a/src/firmware/linux/mod.rs#L122). This contains strictly Rust-friendly types and is meant for Rust use only. Notice that when the actual ioctl is called in the API, the `SnpStatus` type is not used.

2. [SnpPlatformStatus](https://github.com/virtee/sev/blob/d52c5c56b04c44a00674f4190e2f3083376af72f/src/firmware/types.rs#L180): This type is the FFI-friendly type that is presented to the kernel in the ioctl call (it's merely a Rust version of the [sev_user_data_snp_status](https://github.com/AMDESE/linux/blob/6d4469b86f90e67119ff110230857788a0d9dbd0/include/uapi/linux/psp-sev.h#L170) type provided by the kernel [although a seemingly outdated one in the latest SNP kernels]).

In the [snp_platform_status API](https://github.com/virtee/sev/blob/2e54dddffb3ee7cafc805bca203c1136cf41648a/src/firmware/linux/mod.rs#L122), an SnpPlatformStatus is initialized and then passed to the kernel. After this ioctl, the `SnpStatus` struct is built from the data of the `SnpPlatformStatus` struct. The Rust-friendly `SnpStatus` is what is returned from the API.

At some point, Rust-friendly types will need to be built when interacting with the kernel via API's built from these kernel types.

For example, Dov mentioned above that types exposed to the kernel represent a pointer as `u64`. This is apparent here even further, as I wonder how a Rust `Option` is represented in FFI/C-code? A Rust-friendly type could have an `Option` member easily, but types that are exposed to kernel have more stringent rules to avoid errors.

Compilers will give an error with FFI interaction and `Option`:
`warning: block uses type std::option::Option<EXAMPLE> which is not FFI-safe: enum has no representation hint`

Reply:
Created a generic `SnpExtConfig` Rust-friendly struct and set the Get/Set structs to use the u64. I will add more to this as I add implementations.